### PR TITLE
internal/k8s: adopt the RESTMapper interface

### PR DESCRIPTION
**Pull request by jpeach**
The RESTMapper interface wraps up most of the policy around dealing
with detecting API resources and versions. It also has convenience API
for converting between Kind and Resource, which is useful since the
controller-runtime cache deals in kinds, and Contour deals in resources.

This updates #2683.

Signed-off-by: James Peach <jpeach@vmware.com>

### CODE REVIEW

**Comment by jpeach** [comment]
Draft at least until I figure out the right version of controller-runtime.

**Comment by jpeach** [comment]
> Draft at least until I figure out the right version of controller-runtime.

Updated to the latest controller-runtime release, but let's still hold off merging this until after Contour 1.8 releases.

**Review by youngnick — APPROVED** [review]
LGTM

File: examples/contour/01-crds.yaml [hunk 1]
```diff
@@ -41,67 +41,46 @@ spec:
       description: HTTPProxy is an Ingress CRD specification.
       properties:
         apiVersion:
-          description: 'APIVersion defines the versioned schema of this representation
-            of an object. Servers should convert recognized schemas to the latest
-            internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources'
+          description: 'APIVersion defines the versioned schema of this representation of an object. Servers should convert recognized schemas to the latest internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources'
```
Comment (stevesloka) [inline examples/contour/01-crds.yaml:44]: Why did all the crd specs change?

**Comment by jpeach** [comment]
> This looks fine on the top, but what is it solving?

It's a baby step on the path towards controller-runtime. The controller-runtime cache API wants kinds but we always use resources. The RESTMapper interface has convenience functions to do the conversion, which will be useful when we plug it in.

> The code looks the same pretty much just a different client. Also a small nit on why all the CRD descriptions changed.

Yeh, it just seems to be string literal formatting. Probably because the Kubernetes dependencies got a minor version bump?

**Comment by stevesloka** [comment]
> Probably because the Kubernetes dependencies got a minor version bump?

:shakes_fist: at k8s deps. =)

**Pull request merged**